crypto/x509/x_attrib.c

```diff
@@ -56,3 +57,214 @@ X509_ATTRIBUTE *X509_ATTRIBUTE_create(int nid, int atrtype, void *value)
+                 av->value.octet_string->length);
+
+    case V_ASN1_NULL:
+        return BIO_printf(out, "%*sNULL", indent, "");
+
+    case V_ASN1_OBJECT:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        return print_oid(out, av->value.object);
+    
+    /* ObjectDescriptor is an IMPLICIT GraphicString, but GeneralString is a
+    superset supported by OpenSSL, so we will use that anywhere a GraphicString
+    is needed here. */
+    case V_ASN1_GENERALSTRING:
+    case V_ASN1_GRAPHICSTRING:
+    case V_ASN1_OBJECT_DESCRIPTOR:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.generalstring->length,
+                          av->value.generalstring->data);
+
+    /* EXTERNAL would go here. */
+    /* EMBEDDED PDV would go here. */
+
+    case V_ASN1_UTF8STRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.utf8string->length,
+                          av->value.utf8string->data);
+
+    case V_ASN1_REAL:
+        return BIO_printf(out, "%*sREAL", indent, "");
+
+    /* RELATIVE-OID would go here. */
+    /* TIME would go here. */
+
+    case V_ASN1_SEQUENCE:
+        return ASN1_parse_dump(out, av->value.sequence->data,
+                        av->value.sequence->length, indent, 1);
+
+    case V_ASN1_SET:
+        return ASN1_parse_dump(out, av->value.set->data,
+                av->value.set->length, indent, 1);
+
+    /*
+        UTCTime ::= [UNIVERSAL 23] IMPLICIT VisibleString
+        GeneralizedTime ::= [UNIVERSAL 24] IMPLICIT VisibleString
+        VisibleString is a superset for NumericString, so it will work for that.
+    */
+    case V_ASN1_VISIBLESTRING:
+    case V_ASN1_UTCTIME:
+    case V_ASN1_GENERALIZEDTIME:
+    case V_ASN1_NUMERICSTRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.visiblestring->length,
+                          av->value.visiblestring->data);
+
+    case V_ASN1_PRINTABLESTRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.printablestring->length,
+                          av->value.printablestring->data);
+
+    case V_ASN1_T61STRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.t61string->length,
+                          av->value.t61string->data);
+
+    case V_ASN1_IA5STRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.ia5string->length,
+                          av->value.ia5string->data);
+
+    /* UniversalString would go here. */
+    /* CHARACTER STRING would go here. */
+    /* BMPString would go here. */
+    /* DATE would go here. */
+    /* TIME-OF-DAY would go here. */
+    /* DATE-TIME would go here. */
+    /* DURATION would go here. */
+    /* OID-IRI would go here. */
+    /* RELATIVE-OID-IRI would go here. */
```

Comment:
I actually wanted to ask about this. I figured this would be a contentious matter.

1. OpenSSL doesn't have any concept of `DATE`, `TIME-OF-DAY`, `DATE-TIME`, `DURATION`, `OID-IRI` and `RELATIVE-OID-IRI`. These were added as universal types in newer versions of ASN.1. Those types all represented as some sort of string, so I would think if OpenSSL supports those eventually, it will represent them as strings as well. (I am willing to make this PR, by the way. That should be simple enough, assuming we don't care to do any validation on them!). AFAIK, these are not currently used in any extensions or attributes published to date.
2. `UniversalString` and `BMPString` are capable of representing Unicode characters. How does OpenSSL perform output? Can I assume the terminal takes UTF-8 output? It seems like `UTF8String`s are written to `stdout` elsewhere in OpenSSL.
3. `CHARACTER STRING`, `EXTERNAL`, and `EMBEDDED PDV` are constructed types that are rarely used at all. I was thinking that, if I produced an ASN.1 dump for these, it would not be obvious how to differentiate these from `SET` or `SEQUENCE`, but then I realized that this is also true between `SET` and `SEQUENCE`. Should I prefix the ASN.1 dumps with something like `SET:` or `SEQUENCE:`?

Reply:
> Can I assume the terminal takes UTF-8 output?

Yes, that is what other places in OpenSSL assume, so IMO it can be assumed here as well.

Reply:
Anyway I am OK with postponing this.

Reply:
Yeah, I looked into implementing UniversalString and BMPString conversion to UTF-8, and I think that is a little risky, so I'd like to do that in a separate PR so it gets a lot more scrutiny.  And the other types aren't implemented in OpenSSL yet, which I would also like to do in a separate PR. I'll mark this as resolved, then.